File: src/BinarySearch/LC33SearchInRotatedSortedArray.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {boolean}
 */
var search = function (nums, target) {
  let pivot = -1;
  let sp = 0;
  let ep = nums.length - 1;

  // Find pivot
  while (sp <= ep) {
    let mid = Math.trunc((sp + ep) / 2);

    if (mid < ep && nums[mid] > nums[mid + 1]) {
      pivot = mid;
      break;
    } else if (mid > sp && nums[mid] < nums[mid - 1]) {
      pivot = mid - 1;
      break;
    } else if (nums[mid] > nums[0]) {
      sp = mid + 1;
    } else {
      ep = mid - 1;
    }
  }
  if (pivot === -1) pivot = ep;

  // Search in sub-arrays
  if (nums[pivot] === target) {
    return pivot;
  }
  (sp = 0), (ep = pivot);
  while (sp <= ep) {
    let mid = Math.trunc((sp + ep) / 2);

    if (target < nums[mid]) {
      ep = mid - 1;
    } else if (target > nums[mid]) {
      sp = mid + 1;
    } else {
      return mid;
    }
  }
  (sp = pivot + 1), (ep = nums.length - 1);
  while (sp <= ep) {
    let mid = Math.trunc((sp + ep) / 2);

    if (target < nums[mid]) {
      ep = mid - 1;
    } else if (target > nums[mid]) {
      sp = mid + 1;
    } else {
      return mid;
    }
  }

  return -1;
};

console.log(search([2, 5, 6, 0, 0, 1, 2], 3));
console.log(search([2, 5, 6, 0, 0, 1, 2], 0));
console.log(search([1, 0, 1, 1, 1], 0));
console.log(search([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1], 2));
console.log(search([1, 3, 5], 0));

File: src/BinarySearch/LC33SearchInRotatedSortedArray.java
package BinarySearch;

public class LC33SearchInRotatedSortedArray {
    public static void main(String[] args){
        int[] numbers = {4, 5, 6, 0, 1, 2, 3};
        System.out.println(search(numbers, 5));
    }

    public static int search(int[] numbers, int target) {
        // find pivot
        int ps = 0;
        int pe = numbers.length - 1;
        int pm = (ps + pe) / 2;

        while (ps < pe) {
            pm = (ps + pe) / 2;
            if (numbers[pm] > numbers[pm+1]) {
                break;
            }
            // pm in LH
            if (numbers[pm] > numbers[pe]) {
                ps = pm + 1;
            }
            // pm in RH
            else if (numbers[pm] <= numbers[pe]) {
                pe = pm;
            }
        }

        // if target >= first elem = search first half
        if (target >= numbers[0]) {
            return binSearch(numbers, target, 0, pm);
        }
        // if target < first elem = search last half
        else {
            return binSearch(numbers, target, pm, numbers.length - 1);
        }
    }

    public static int binSearch(int[] numbers, int target, int start, int end) {
        int pm;

        while (end >= start) {
            pm = start + ((end - start) / 2);
            if (target == numbers[pm])
                return pm;

            if (target > numbers[pm]) {
                start = pm + 1;
            } else {
                end = pm - 1;
            }
        }

        return -1;
    }
}